# 1. Diferentes tipos de problemas de ciências de dados

## 1.1. Problema de regressão

Estamos prevendo uma valor numérico em uma escala contínua. 

## 1.2. Problema de classificação

Previsão qualitativa, por exemplos, respostas sim e não.

## 1.3. Aprendizado Supervisionado

É um tipo de prblema que  depende de dados rotulados. 

## 1.4. Aprendizado Não Supervisionado

Tentamos encontrar algum tipo de estrutura em um dataset que nem sempre tem rótulos.

# 2. Estudo de Caso

Dados disponíveis em: https://github.com/TrainingByPackt/Data-Science-Projects-with-Python/tree/master/Data

# 3. Carregando os pacotes

In [2]:
import pandas as pd

# 4. Carregando a base de dados

In [3]:
df = pd.read_excel('data/default_of_credit_card_clients__courseware_version_1_21_19.xls')

## 4.1. Métodos para usar com uma data frame

- `.shape()`
- `.info()`
- `.head()`

In [4]:
df.shape()

TypeError: 'tuple' object is not callable

In [5]:
df.shape #numero de linhas e colunas

(30000, 25)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   ID                          30000 non-null  object
 1   LIMIT_BAL                   30000 non-null  int64 
 2   SEX                         30000 non-null  int64 
 3   EDUCATION                   30000 non-null  int64 
 4   MARRIAGE                    30000 non-null  int64 
 5   AGE                         30000 non-null  int64 
 6   PAY_1                       30000 non-null  object
 7   PAY_2                       30000 non-null  int64 
 8   PAY_3                       30000 non-null  int64 
 9   PAY_4                       30000 non-null  int64 
 10  PAY_5                       30000 non-null  int64 
 11  PAY_6                       30000 non-null  int64 
 12  BILL_AMT1                   30000 non-null  int64 
 13  BILL_AMT2                   30000 non-null  in

In [7]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  PAY_2  \
0  798fc410-45c1      20000    2          2         1   24     2      2   
1  8a8c8f3b-8eb4     120000    2          2         2   26    -1      2   
2  85698822-43f5      90000    2          2         2   34     0      0   
3  0737c11b-be42      50000    2          2         1   37     0      0   
4  3b7f77cc-dbc0      50000    1          2         1   57    -1      0   

   PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0     -1     -1  ...          0          0          0         0       689   
1      0      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4     -1      0  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 25 columns]

O dicionário de dados pode ser encontrado aqui: https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients

# 5. O problema da empresa

- Nosso cliente é uma empresa de cartão de crédito.
- A base de dados contém dados demográficos de 30.000 titulares de contas correntes.

## 5.1. Objetivo

Desenvolver um modelo que preveja se uma conta ficará inadimplente no próximo mês, de acordo com dados demográficos e históricos.

# 6. Etapas da Exploração de dados

## 6.1. Quantas colunas os dados contêm?

Temos 25 colunas ou variáveis (features)

## 6.2. Quantas linhas os dados contêm?

Temos 30.000 observações (ou linhas).

## 6.3. Examine os nomes das colunas usando o método `.columns`.

In [8]:
df.columns.to_list()

['ID',
 'LIMIT_BAL',
 'SEX',
 'EDUCATION',
 'MARRIAGE',
 'AGE',
 'PAY_1',
 'PAY_2',
 'PAY_3',
 'PAY_4',
 'PAY_5',
 'PAY_6',
 'BILL_AMT1',
 'BILL_AMT2',
 'BILL_AMT3',
 'BILL_AMT4',
 'BILL_AMT5',
 'BILL_AMT6',
 'PAY_AMT1',
 'PAY_AMT2',
 'PAY_AMT3',
 'PAY_AMT4',
 'PAY_AMT5',
 'PAY_AMT6',
 'default payment next month']

## 6.4. Selecione  a coluna alvo `ID` e conte os valores exclusivos usando o método `.nunique()`

In [9]:
df['ID'].nunique()

29687

Temos 29687 IDs unicos e 313 IDs que estão repetidos.

## 6.5. Há valores de ID duplicados no dataset. Vamos criar uma variável chamada `id_counts()` para contar quandtas vezes cada ID se repete.

In [4]:
id_counts = df['ID'].value_counts()
id_counts.head()

428b26a3-dca2    2
e50d8395-da32    2
d5905747-8aef    2
df3d13be-9c13    2
4e2380e6-a8cf    2
Name: ID, dtype: int64

## 6.6. Aplique o método `value_counts()` em `id_counts` para ver o número de entradas duplicadas.

# 7. Máscaras booleanas

In [10]:
type(df['ID'])

pandas.core.series.Series

Exemplo de uso de máscara booleana:

In [11]:
df['AGE'] >= 30

0        False
1        False
2         True
3         True
4         True
         ...  
29995     True
29996     True
29997     True
29998     True
29999     True
Name: AGE, Length: 30000, dtype: bool

In [12]:
import nupy as np

ModuleNotFoundError: No module named 'nupy'

In [13]:
import numpy as np